Fix all errors and warnings in CMake, VS2017RC and Xcode.

gonsolo opened the pull request:
This fixes a warning about truncation from size_t (64 bits on 64 bit Windows) to int (32 bits on 64 bit Windows).

sergiud commented on src/demangle.cc:
```diff
@@ -340,7 +340,7 @@ static void MaybeAppendWithLength(State *state, const char * const str,
 static bool MaybeAppend(State *state, const char * const str) {
   if (state->append) {
-    int length = StrLen(str);
+    auto length = StrLen(str);
```
This introduces an unconditional dependency to C++11. However, glog is a C++98/03 library.

gonsolo replied:
Changed. Is it ok now?

gonsolo commented:
Changed from auto to size_t.

gonsolo commented:
Now all errors and warnings from CMake, Visual Studio 2017 RC and Xcode 8.2.1 are fixed. Please pull.

sergiud commented on CMakeLists.txt:
```diff
@@ -26,8 +26,8 @@ set (CPACK_PACKAGE_VERSION_MINOR ${GLOG_MINOR_VERSION})
-option (WITH_GFLAGS "Use gflags" ON)
-option (WITH_THREADS "Enable multithreading support" ON)
+option (WITH_GFLAGS "Use gflags" OFF)
+option (WITH_THREADS "Enable multithreading support" OFF)
```
Please do not change the defaults. Gflags and Threads are both optional features, even if the `WITH_*` options are set to `ON`. The messages you see during generation are not errors, but warnings.

gonsolo replied:
Have you a suggestion for a error- and warning-free default checkout?

sergiud replied:
You can simply ignore the warnings about missing `gflags-config.cmake`. There are also CMake means for disabling packages using the command-line. You'll have to check the CMake documentation.

Are there any other warnings/errors that prevent generation? A log output would be helpful.

sergiud commented on src/logging.cc:
```diff
@@ -191,7 +191,7 @@ GLOG_DEFINE_string(log_backtrace_at, "",
 #define ssize_t SSIZE_T
 #endif
-static ssize_t pread(int fd, void* buf, size_t count, off_t offset) {
+static ssize_t pread(int fd, void* buf, unsigned int count, off_t offset) {
```
The changes from `unsigned int` to `size_t` do not look right.

gonsolo replied:
Ok. I changed it to size_t again and added a cast for read which uses unsigned int.

sergiud commented on src/logging.cc:
```diff
@@ -1436,7 +1436,7 @@ void LogMessage::SendToLog() EXCLUSIVE_LOCKS_REQUIRED(log_mutex) {
 
     const char* message = "*** Check failure stack trace: ***\n";
-    if (write(STDERR_FILENO, message, strlen(message)) < 0) {
+    if (write(STDERR_FILENO, message, (unsigned int)strlen(message)) < 0) {
```
Why did you change the parameter type if you still need to perform a type cast?

gonsolo replied:
Changed to #pragma warning( disable: 4267 ) since on Windows strlen returns size_t while write uses unsigned int.

sergiud commented on src/logging.cc:
```diff
@@ -2019,6 +2019,9 @@ LogMessageFatal::LogMessageFatal(const char* file, int line,
 
+#ifdef _MSC_VER
+# pragma warning(disable: 4722)
+#endif
```
Use push/pop to restore the warnings.

gonsolo replied:
Done.

gonsolo commented:
I addressed all comments except the one about disabling WITH_ defaults. How can I make this warning go away?

sergiud commented:
@gonsolo See my previous comment. The warning (which is of informative nature) can be simply ignored.

gonsolo commented:
Ok. Changed back. I addressed all other comments. Is there anything still
to do before you can pull?

And thanks for reviewing. :)

On Wed, Jan 18, 2017 at 10:50 AM, Sergiu Deitsch <notifications@github.com>
wrote:

> @gonsolo <https://github.com/gonsolo> See my previous comment. The
> warning (which is of informative nature) can be simply ignored.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/glog/pull/156#issuecomment-273431002>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB8npGW2mT8mggH5_SLNSBpsWIgNbmGwks5rTeBwgaJpZM4LkqOe>
> .
>



-- 
g

sergiud commented:
Just squash the commits into one using `git rebase -i`. There are just too many.

gonsolo commented:
Done.

On Wed, Jan 18, 2017 at 11:25 AM, Sergiu Deitsch <notifications@github.com>
wrote:

> Just squash the commits into one using git rebase -i. There are just too
> many.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/glog/pull/156#issuecomment-273438895>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB8npEI5kUV8tmZoJ7fUBAFhD1dSaEmYks5rTeh9gaJpZM4LkqOe>
> .
>



-- 
g

sergiud commented on src/logging.cc:
```diff
@@ -207,7 +207,7 @@ static ssize_t pread(int fd, void* buf, size_t count, off_t offset) {
 
 #ifndef HAVE_PWRITE
-static ssize_t pwrite(int fd, void* buf, size_t count, off_t offset) {
+static ssize_t pwrite(int fd, void* buf, unsigned int count, off_t offset) {
```
`size_t` is the correct parameter type.

sergiud replied:
No change here?

sergiud commented on src/logging.cc:
```diff
@@ -628,7 +628,7 @@ inline void LogDestination::RemoveLogSink(LogSink *destination) {
   // This doesn't keep the sinks in order, but who cares?
   if (sinks_) {
-    for (int i = sinks_->size() - 1; i >= 0; i--) {
+    for (int i = (int)sinks_->size() - 1; i >= 0; i--) {
```
`ssize_t` or `ptrdiff_t` instead of the type cast to `int` makes more sense to avoid overflows.

gonsolo replied:
ssize_t makes no sense since the comparison >=0 would be useless. Using ptrdiff_t.

sergiud replied:
I don't understand. `ssize_t` is the signed counterpart of `size_t`.

sergiud replied:
But `ptrdiff_t` is fine. `ptrdiff_t` is actually a better choice, since the type is standardized in C++, whereas `ssize_t` is not.

sergiud commented on src/logging.cc:
```diff
@@ -786,7 +786,7 @@ inline void LogDestination::LogToSinks(LogSeverity severity,
   ReaderMutexLock l(&sink_mutex_);
   if (sinks_) {
-    for (int i = sinks_->size() - 1; i >= 0; i--) {
+    for (int i = (int)sinks_->size() - 1; i >= 0; i--) {
```
Same as before.

sergiud commented on src/logging.cc:
```diff
@@ -796,7 +796,7 @@ inline void LogDestination::LogToSinks(LogSeverity severity,
   ReaderMutexLock l(&sink_mutex_);
   if (sinks_) {
-    for (int i = sinks_->size() - 1; i >= 0; i--) {
+    for (int i = (int)sinks_->size() - 1; i >= 0; i--) {
```
Here also.

sergiud commented on src/logging_unittest.cc:
```diff
@@ -251,8 +251,8 @@ void TestLogging(bool check_counts) {
-    LOG_EVERY_N(ERROR, 3) << "Log every 3, iteration " << COUNTER << endl;
-    LOG_EVERY_N(ERROR, 4) << "Log every 4, iteration " << COUNTER << endl;
+    //LOG_EVERY_N(ERROR, 3) << "Log every 3, iteration " << COUNTER << endl;
+    //LOG_EVERY_N(ERROR, 4) << "Log every 4, iteration " << COUNTER << endl;
```
Why did you comment out these lines?

gonsolo replied:
It gives me a warning on Xcode about "Unused typedef INVALID_REQUESTED_LOG_SEVERITY" which I am unable to find anywhere in the source.

sergiud replied:
Maybe using `__attribute__((unused))` can help? The typedef is likely generated using macros. Try searching for a small part of typedef's name.

sergiud commented on src/raw_logging.cc:
```diff
@@ -150,7 +150,7 @@ void RawLog__(LogSeverity severity, const char* file, int line,
   // libc (to side-step any libc interception).
   // We write just once to avoid races with other invocations of RawLog__.
-  safe_write(STDERR_FILENO, buffer, strlen(buffer));
+  safe_write(STDERR_FILENO, buffer, (unsigned int)strlen(buffer));
```
You should avoid the type cast.

gonsolo replied:
Ok.

gonsolo commented:
Oh, I misread ssize_t for size_t. Sorry.

I am using ptrdiff_t now. Do you want me to change that to ssize_t?

On Wed, Jan 18, 2017 at 12:47 PM, Sergiu Deitsch <notifications@github.com>
wrote:

> *@sergiud* commented on this pull request.
> ------------------------------
>
> In src/logging.cc <https://github.com/google/glog/pull/156>:
>
> > @@ -628,7 +628,7 @@ inline void LogDestination::RemoveLogSink(LogSink *destination) {
>    MutexLock l(&sink_mutex_);
>    // This doesn't keep the sinks in order, but who cares?
>    if (sinks_) {
> -    for (int i = sinks_->size() - 1; i >= 0; i--) {
> +    for (int i = (int)sinks_->size() - 1; i >= 0; i--) {
>
> I don't understand. ssize_t is the signed counterpart of size_t.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/glog/pull/156>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB8npL1Jp0OLPC6qOlJJeTpAHgo_YYS8ks5rTfvXgaJpZM4LkqOe>
> .
>



-- 
g

gonsolo commented:
Ok. I see. Thanks.

On Wed, Jan 18, 2017 at 12:51 PM, Gonsolo <gonsolo@gmail.com> wrote:

> Oh, I misread ssize_t for size_t. Sorry.
>
> I am using ptrdiff_t now. Do you want me to change that to ssize_t?
>
> On Wed, Jan 18, 2017 at 12:47 PM, Sergiu Deitsch <notifications@github.com
> > wrote:
>
>> *@sergiud* commented on this pull request.
>> ------------------------------
>>
>> In src/logging.cc <https://github.com/google/glog/pull/156>:
>>
>> > @@ -628,7 +628,7 @@ inline void LogDestination::RemoveLogSink(LogSink *destination) {
>>    MutexLock l(&sink_mutex_);
>>    // This doesn't keep the sinks in order, but who cares?
>>    if (sinks_) {
>> -    for (int i = sinks_->size() - 1; i >= 0; i--) {
>> +    for (int i = (int)sinks_->size() - 1; i >= 0; i--) {
>>
>> I don't understand. ssize_t is the signed counterpart of size_t.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/google/glog/pull/156>, or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AB8npL1Jp0OLPC6qOlJJeTpAHgo_YYS8ks5rTfvXgaJpZM4LkqOe>
>> .
>>
>
>
>
> --
> g
>



-- 
g

gonsolo commented:
Ok, found it. The typedef is defined in a macro and subsequently not used.
Removed it and pushed.

On Wed, Jan 18, 2017 at 12:52 PM, Sergiu Deitsch <notifications@github.com>
wrote:

> *@sergiud* commented on this pull request.
> ------------------------------
>
> In src/logging_unittest.cc <https://github.com/google/glog/pull/156>:
>
> > @@ -251,8 +251,8 @@ void TestLogging(bool check_counts) {
>      PLOG_EVERY_N(ERROR, 2) << "Plog every 2, iteration " << COUNTER;
>      errno = old_errno;
>
> -    LOG_EVERY_N(ERROR, 3) << "Log every 3, iteration " << COUNTER << endl;
> -    LOG_EVERY_N(ERROR, 4) << "Log every 4, iteration " << COUNTER << endl;
> +    //LOG_EVERY_N(ERROR, 3) << "Log every 3, iteration " << COUNTER << endl;
> +    //LOG_EVERY_N(ERROR, 4) << "Log every 4, iteration " << COUNTER << endl;
>
> Maybe using __attribute__((unused)) can help? The typedef is likely
> generated using macros. Try searching for a small part of typedef's name.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/glog/pull/156>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB8npAm5eGxxZNjXaJ4G9cGcJ48nCsP3ks5rTf0ZgaJpZM4LkqOe>
> .
>



-- 
g

gonsolo commented:
Found a even better solution: The typedef is a compile check so it is just
marked unused.

On Wed, Jan 18, 2017 at 1:27 PM, Gonsolo <gonsolo@gmail.com> wrote:

> Ok, found it. The typedef is defined in a macro and subsequently not used.
> Removed it and pushed.
>
> On Wed, Jan 18, 2017 at 12:52 PM, Sergiu Deitsch <notifications@github.com
> > wrote:
>
>> *@sergiud* commented on this pull request.
>> ------------------------------
>>
>> In src/logging_unittest.cc <https://github.com/google/glog/pull/156>:
>>
>> > @@ -251,8 +251,8 @@ void TestLogging(bool check_counts) {
>>      PLOG_EVERY_N(ERROR, 2) << "Plog every 2, iteration " << COUNTER;
>>      errno = old_errno;
>>
>> -    LOG_EVERY_N(ERROR, 3) << "Log every 3, iteration " << COUNTER << endl;
>> -    LOG_EVERY_N(ERROR, 4) << "Log every 4, iteration " << COUNTER << endl;
>> +    //LOG_EVERY_N(ERROR, 3) << "Log every 3, iteration " << COUNTER << endl;
>> +    //LOG_EVERY_N(ERROR, 4) << "Log every 4, iteration " << COUNTER << endl;
>>
>> Maybe using __attribute__((unused)) can help? The typedef is likely
>> generated using macros. Try searching for a small part of typedef's name.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/google/glog/pull/156>, or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AB8npAm5eGxxZNjXaJ4G9cGcJ48nCsP3ks5rTf0ZgaJpZM4LkqOe>
>> .
>>
>
>
>
> --
> g
>



-- 
g

gonsolo commented:
Fixed whitespace and rebased. Should be ready to pull now.

On Wed, Jan 18, 2017 at 2:13 PM, Gonsolo <gonsolo@gmail.com> wrote:

> Found a even better solution: The typedef is a compile check so it is just
> marked unused.
>
> On Wed, Jan 18, 2017 at 1:27 PM, Gonsolo <gonsolo@gmail.com> wrote:
>
>> Ok, found it. The typedef is defined in a macro and subsequently not
>> used. Removed it and pushed.
>>
>> On Wed, Jan 18, 2017 at 12:52 PM, Sergiu Deitsch <
>> notifications@github.com> wrote:
>>
>>> *@sergiud* commented on this pull request.
>>> ------------------------------
>>>
>>> In src/logging_unittest.cc <https://github.com/google/glog/pull/156>:
>>>
>>> > @@ -251,8 +251,8 @@ void TestLogging(bool check_counts) {
>>>      PLOG_EVERY_N(ERROR, 2) << "Plog every 2, iteration " << COUNTER;
>>>      errno = old_errno;
>>>
>>> -    LOG_EVERY_N(ERROR, 3) << "Log every 3, iteration " << COUNTER << endl;
>>> -    LOG_EVERY_N(ERROR, 4) << "Log every 4, iteration " << COUNTER << endl;
>>> +    //LOG_EVERY_N(ERROR, 3) << "Log every 3, iteration " << COUNTER << endl;
>>> +    //LOG_EVERY_N(ERROR, 4) << "Log every 4, iteration " << COUNTER << endl;
>>>
>>> Maybe using __attribute__((unused)) can help? The typedef is likely
>>> generated using macros. Try searching for a small part of typedef's name.
>>>
>>> —
>>> You are receiving this because you were mentioned.
>>> Reply to this email directly, view it on GitHub
>>> <https://github.com/google/glog/pull/156>, or mute the thread
>>> <https://github.com/notifications/unsubscribe-auth/AB8npAm5eGxxZNjXaJ4G9cGcJ48nCsP3ks5rTf0ZgaJpZM4LkqOe>
>>> .
>>>
>>
>>
>>
>> --
>> g
>>
>
>
>
> --
> g
>



-- 
g

sergiud commented on src/windows/glog/logging.h:
```diff
@@ -932,8 +932,14 @@ struct CrashReason;
 bool IsFailureSignalHandlerInstalled();
 }  // namespace glog_internal_namespace_
 
+#ifdef __APPLE__
```
This is not Apple specific. GCC and Clang support the attribute.

sergiud commented on src/logging_unittest.cc:
```diff
@@ -795,7 +795,7 @@ static void TestOneTruncate(const char *path, int64 limit, int64 keep,
   char* buf = new char[buf_size];
   memset(buf, 0, buf_size);
-  CHECK_ERR(read(fd, buf, buf_size));
+  CHECK_ERR(read(fd, buf, (unsigned int)buf_size));
```
Why do you cast here?

sergiud commented on src/logging.cc:
```diff
@@ -786,7 +794,7 @@ inline void LogDestination::LogToSinks(LogSeverity severity,
   ReaderMutexLock l(&sink_mutex_);
   if (sinks_) {
-    for (int i = sinks_->size() - 1; i >= 0; i--) {
+    for (ptrdiff_t i = sinks_->size() - 1; i >= 0; i--) {
```
Just thought about this. You can write

```c++
for (size_t i = sinks_->size() - 1; i < sinks_->size(); --i)
    // ...
```
If `i` overflows, the loop will end. This is also safe, since unsigned integers are guaranteed to wrap.

gonsolo replied:
Ok. Is ptrdiff_t ok though?

sergiud replied:
You don't need `ptrdiff_t` if you use the loop above.

gonsolo replied:
I'm pretty sure I will get a warning about truncation from size_t (sinks_->size()) to int (i). This was the reason to change the loop in the first place.

sergiud replied:
This can be right. `sinks_` is of type `std::vector<LogSink*>` whose `size` method returns a type compatible to `size_t`. So the proposed should be just fine.

sergiud replied:
Or you can use `std::vector<LogSink*>::reverse_iterator`.

sergiud commented on src/logging.cc:
```diff
@@ -796,7 +804,7 @@ inline void LogDestination::LogToSinks(LogSeverity severity,
   ReaderMutexLock l(&sink_mutex_);
   if (sinks_) {
-    for (int i = sinks_->size() - 1; i >= 0; i--) {
+    for (ptrdiff_t i = sinks_->size() - 1; i >= 0; i--) {
```
Same as before.

sergiud commented on src/logging.cc:
```diff
@@ -628,7 +628,7 @@ inline void LogDestination::RemoveLogSink(LogSink *destination) {
   // This doesn't keep the sinks in order, but who cares?
   if (sinks_) {
-    for (int i = sinks_->size() - 1; i >= 0; i--) {
+    for (ptrdiff_t i = sinks_->size() - 1; i >= 0; i--) {
```
See my next comment.

sergiud commented on src/logging.cc:
```diff
@@ -207,13 +207,13 @@ static ssize_t pread(int fd, void* buf, size_t count, off_t offset) {
   if (lseek(fd, offset, SEEK_CUR) == (off_t)-1)
     return -1;
-  ssize_t len = write(fd, buf, count);
+  ssize_t len = write(fd, buf, (unsigned int)count);
```
You should avoid the cast here.

sergiud commented on src/logging.cc:
```diff
@@ -191,13 +191,13 @@ GLOG_DEFINE_string(log_backtrace_at, "",
   if (lseek(fd, offset, SEEK_CUR) == (off_t)-1)
     return -1;
-  ssize_t len = read(fd, buf, count);
+  ssize_t len = read(fd, buf, (unsigned int)count);
```
Why the type cast?

gonsolo commented:
The source problem is that strlen returns size_t but _read and _write use int on Windows. So somewhere it has to be cast. I tried to leave the source mostly as it was now (or changed to size_t) and just disabled the warnings where it happened.

gonsolo commented:
Fixed and rebased again. Next try. :)

gonsolo commented:
>
> This can be right. sinks_ is of type std::vector<LogSink*> whose size
> method returns a type compatible to size_t. So the proposed should be
> just fine.
>

I was talking of the original version where sinks_ returned size_t but it
was assigned to an int.

-- 
g

gonsolo commented:
>
> Or you can use std::vector<LogSink*>::reverse_iterator.
>

Can we just stick with ptrdiff_t? I just wanted to get rid of the warning
and not think of the program logic too much. Changing this could introduce
different semantics or subtle errors.

-- 
g

gonsolo commented:
FWIW, I'm using glog in a bigger project and I just tried to get rid
of all warnings so new warnings in the project will stand out clearly.

-- 
g

sergiud commented:
@gonsolo The way the loop is implemented now, is incorrect. By using my proposal you will fix this and get rid of the corresponding warnings.

gonsolo commented:
>
> @gonsolo <https://github.com/gonsolo> The way the loop is implemented
> now, is incorrect. By using my proposal you will fix this and get rid of
> the corresponding warnings.
>

Old version:
   for (int i = sinks_->size() - 1; i >= 0; i--) {

My version:
   for (ptrdiff_t i = sinks_->size() - 1; i >= 0; i--) {

I don't get it, what's wrong here?

-- 
g

sergiud commented:
@gonsolo Correct me if I'm wrong, but `size()` return value (which is unsigned) is truncated by conversion to `ptrdiff_t` (which is signed).

gonsolo commented:
>
> @gonsolo <https://github.com/gonsolo> Correct me if I'm wrong, but size()
> return value (which is unsigned) is truncated by conversion to ptrdiff_t
> (which is signed).
>

Yes. The old version truncated to 32 bits signed (int) and the new version
to 64 bit signed (signed). VS warns about the former but not the latter. So
it's progress. :)
Anyway, I changed it to reverse iterators.

There is one pop_back in a loop but I guess it's ok since it only
invalidates the last element, right?

-- 
g

gonsolo commented:
Anything else I can improve?

sergiud commented:
@gonsolo Could you please resolve the conflicts?

gonsolo commented:
No, it's been nearly three years and I stopped using glog.
